python/cudf/cudf/core/dataframe.py

```diff
@@ -6431,71 +6431,314 @@ def to_orc(
-                    else self._data[c]
-                )
-                for c in self._data
-            }
+        # Compute the column indices that serves as the input for
+        # `interleave_columns`
+        column_idx_df = pd.DataFrame(
+            data=range(len(self._data)), index=named_levels
         )
 
-        result = Series._from_data(
-            {
-                None: libcudf.reshape.interleave_columns(
-                    [*homogenized._columns]
+        column_indices: list[list[int]] = []
+        unnamed_level_values = None
+
+        def unnamed_group_generator():
+            if len(unnamed_levels_indices) == 0:
+                yield column_idx_df
+            else:
+                nonlocal unnamed_level_values
+                unnamed_level_values = [
+                    column_name_idx.get_level_values(lv)
+                    for lv in unnamed_levels_indices
+                ]
+                unnamed_level_values = pd.MultiIndex.from_arrays(
+                    unnamed_level_values
                 )
-            },
-            index=new_index,
-        )
+                for _, grpdf in column_idx_df.groupby(by=unnamed_level_values):
+                    yield grpdf
+
+        for df in unnamed_group_generator():
+            if len(unnamed_levels_indices) == 0:
+                grpdf_aligned = df
+            else:
+                # When stacking part of the levels, some combinations of keys
+                # may not present in this group but can present in others.
+                # Reindexing with the globally computed `unique_named_levels`
+                # assigns -1 to these key combinations, representing an
+                # all-null column that is used in the subsequent libcudf call.
+                grpdf_aligned = df.reindex(
+                    unique_named_levels, axis=0, fill_value=-1
+                )
+            grpdf_aligned.sort_index(inplace=True)  # this needed?
+            indices = []
+            for label in grpdf_aligned.index:
+                (column_idx,) = grpdf_aligned.loc[label]
+                indices.append(int(column_idx))
+            column_indices.append(indices)
+
+        # For each of the group constructed from the unnamed levels,
+        # invoke `interleave_columns` to stack the values.
+        stacked = []
+        all_nans = None  # lazily created when needed
+        for column_idx in column_indices:
+            columns: list(column | None) = []
+
+            # Collect columns based on indices, append None for -1 indices.
+            for i in column_idx:
+                if i == -1:
+                    columns.append(None)
+                else:
+                    columns.append(self._data.select_by_index(i).columns[0])
+
+            # Collect datatypes and cast columns as that type
+            common_type = np.result_type(
+                *[col.dtype for col in columns if col is not None]
```

Comment:
```suggestion
                *(col.dtype for col in columns if col is not None)
```
No need to cons the list to immediately splat it.

Reply:
Nice, didn't thought that list comprehension is also cons.